javaslang/src/main/java/javaslang/collection/Arrays2.java

```diff
@@ -0,0 +1,131 @@
+/*     / \____  _    _  ____   ______  / \ ____  __    _______
+ *    /  /    \/ \  / \/    \ /  /\__\/  //    \/  \  //  /\__\   JΛVΛSLΛNG
+ *  _/  /  /\  \  \/  /  /\  \\__\\  \  //  /\  \ /\\/ \ /__\ \   Copyright 2014-2016 Javaslang, http://javaslang.io
+ * /___/\_/  \_/\____/\_/  \_/\__\/__/\__\_/  \_//  \__/\_____/   Licensed under the Apache License, Version 2.0
+ */
+package javaslang.collection;
+
+import java.util.Arrays;
+import java.util.Random;
+
+/**
+ * Internal class, containing helpers.
+ *
+ * @author Pap Lőrinc
+ * @since 3.0.0
+ */
+@SuppressWarnings({"unchecked", "SuspiciousArrayCast"})
+final class Arrays2 { // TODO reuse these in `Array` also
+    static final Object[] EMPTY = {};
+
+    static <T> T[] emptyArray()                  { return (T[]) EMPTY; }
+    static boolean isNullOrEmpty(Object[] array) { return (array == null) || (array.length == 0); }
+
+    public static Object get(Object[] array, int i) {
+        return (isNullOrEmpty(array) || (i < 0) || (i >= array.length))
+               ? null
+               : array[i];
+    }
+
+    /** Group an array into many equal sized sub-arrays */
+    static Object[] grouped(Object[] array, int length, int size) {
+        Object[] results = new Object[Math.min(size, length)];
+        System.arraycopy(array, 0, results, 0, results.length);
+
+        if (results.length < array.length) {
+            final Object[] parentArray = new Object[1 + ((length - 1) / size)];
+            parentArray[0] = results;
+
+            Object[] next = new Object[Math.min(size, length - size)];
+            System.arraycopy(array, results.length, next, 0, next.length);
+            parentArray[1] = next;
+
+            for (int start = results.length + next.length, i = 2; start < array.length; i++) {
+                next = new Object[Math.min(size, length - (i * size))];
+                System.arraycopy(array, start, next, 0, next.length);
+                parentArray[i] = next;
+                start += next.length;
+            }
+
+            results = parentArray;
+        }
+
+        return results;
+    }
+
+    /** Store the content of an iterable in an array */
+    static <T> Object[] asArray(java.util.Iterator<?> it, int length) {
+        final Object[] array = new Object[length];
+        for (int i = 0; i < length; i++) {
+            array[i] = it.next();
+        }
+        return array;
+    }
+
+    static <T> Iterator<T> iterator(final T[] array, int offset) {
+        return new Iterator<T>() { /* it's faster than Iterator.of(root) */
+            private int index = offset;
+
+            @Override
+            public boolean hasNext() { return index < array.length; }
+
+            @Override
+            public T next() { return array[index++]; }
+        };
+    }
+
+    static <T> T[] inject(Object arrayObject, int index, T element) {
```

Comment:
It's somewhat counter-intuitive to me that this method covers three different use cases:
1. injecting to an empty or null array produces a new single-element array, ignoring `index`;
2. injecting with a positive index "inserts" the elem to the `index` position;
3. injecting with a negative index shifts the whole array and inserts the to the beginning.